torsiondrive/qm_engine.py

```diff
@@ -327,6 +330,114 @@ def load_native_output(self, filename='output.dat'):
+            assert '--opt' in cmd, "comment line should contain --opt command to use native opt"
+
+        if '--parallel' in cmd:
+            self.core = int(cmd[cmd.index('--parallel')+len('--parallel'):].split()[0])
+            pass
+        elif '-P' in cmd:
+            self.core = int(cmd[cmd.index('-P')+len('-P'):].split()[0])
+        else:
+            # Run xTB in single core
+            self.core = 1
+            cmd += ' --parallel 1 '
+
+        self.xTB_args = cmd
+        # here self.M can be and will be overwritten by external functions
+        self.M = Molecule()
+        self.M.elem = elems
+        self.M.xyzs = [np.array(coords, dtype=float)]
+        self.M.build_topology()
+
+    def write_input(self, filename='coords'):
+        """ Write output based on self.M, using only geometry of the first frame """
+        # Write the coordinate file
+        with open(filename + '.xyz', 'w') as outfile:
+            outfile.write('{}\n\n'.format(len(self.M.elem)))
+            for e, c in zip(self.M.elem, self.M.xyzs[0]):
+                outfile.write("{:<7} {:<13.7f} {:<13.7f} {:<13.7f}\n".format(e, c[0], c[1], c[2]))
+        with open(filename + '.inp', 'w') as outfile:
+            outfile.write(self.opt_restrain)
+
+    def optimize_native(self):
+        """ run the constrained optimization using native opt, in 2 steps:
+        1. write a optimization job input file.
+        2. run the job
+        """
+        if self.extra_constraints is not None:
+            raise RuntimeError('extra constraints not supported in xTB native optimizations')
+        # add the optking command
+        self.opt_restrain = '$constrain\n'
+        self.opt_restrain += 'force constant=15.0\n' # Unit Hatree/rad^2 to obtain a STD of 0.5 degree
+        for d1, d2, d3, d4, v in self.dihedral_idx_values:
+            # xTB use atom index starting from 1
+            self.opt_restrain += 'dihedral: {D1}, {D2}, {D3}, {D4}, {v}\n'.format(D1=d1+1, D2=d2+1, D3=d3+1, D4=d4+1, v=v)
+        self.opt_restrain += '$end\n'
+        # write input file
+        self.write_input('input')
+        # run the job
+        self.run('xtb input.xyz {cmd} --input input.inp'.format(cmd=self.xTB_args),
+                 input_files=['input.inp', 'input.xyz'], output_files=['xtbopt.xyz'])
+
+    def load_native_output(self, filename='xtbopt.xyz'):
+        """ Load the optimized geometry and energy into a new molecule object and return """
+        final_energy, elems, coords = None, [], []
+        with open(filename) as outfile:
+            n_atoms = int(outfile.readline())
+            comment = outfile.readline()
+            for line in range(n_atoms):
+                elem, x, y, z = outfile.readline().split()
+                elems.append(elem)
+                coords.append((x, y, z))
+        try:
+            # Also checks if right number of fields are present
+            _, final_energy, _, _, _, _, _ = comment.split()
```

Comment:
This output was changed in the past and is not guaranteed to stay in the current format. The first real number being the energy however is something you can rely on